lib/puppet/application/lookup.rb

```diff
@@ -373,38 +373,34 @@ def generate_scope
     end
 
     unless node.is_a?(Puppet::Node) # to allow unit tests to pass a node instance
-      facts = retrieve_node_facts(node, given_facts) 
-      if Puppet.settings.set_by_cli?('environment')
-        node = Puppet::Node.new(node, :classes => nil, :parameters => nil, :facts => facts, :environment => Puppet.settings.value('environment'))
-      else
-        ni = Puppet::Node.indirection
-        tc = ni.terminus_class
-        if options[:compile]
-          if tc == :plain
-            node = ni.find(node, facts: facts)
-          else
-            begin
-              service = Puppet.runtime[:http]
-              session = service.create_session
-              cert = session.route_to(:ca)
-
-              _, x509 = cert.get_certificate(node)
-              cert = OpenSSL::X509::Certificate.new(x509)
-              Puppet::SSL::Oids.register_puppet_oids
-              trusted = Puppet::Context::TrustedInformation.remote(true, facts.values['certname'] || node, Puppet::SSL::Certificate.from_instance(cert))
-              Puppet.override(trusted_information: trusted) do
-                node = ni.find(node, facts: facts)
-              end
-            rescue
-              Puppet.warning _("CA is not available, the operation will continue without using trusted facts.")
+      facts = retrieve_node_facts(node, given_facts)
+      ni = Puppet::Node.indirection
+      tc = ni.terminus_class
+      if options[:compile] && !Puppet.settings.set_by_cli?('environment')
+        if tc == :plain
+          node = ni.find(node, facts: facts)
+        else
+          begin
+            service = Puppet.runtime[:http]
+            session = service.create_session
+            cert = session.route_to(:ca)
+
+            _, x509 = cert.get_certificate(node)
+            cert = OpenSSL::X509::Certificate.new(x509)
+            Puppet::SSL::Oids.register_puppet_oids
+            trusted = Puppet::Context::TrustedInformation.remote(true, facts.values['certname'] || node, Puppet::SSL::Certificate.from_instance(cert))
+            Puppet.override(trusted_information: trusted) do
               node = ni.find(node, facts: facts)
             end
+          rescue
+            Puppet.warning _("CA is not available, the operation will continue without using trusted facts.")
+            node = ni.find(node, facts: facts)
           end
-        else
-          ni.terminus_class = :plain
-          node = ni.find(node, facts: facts)
-          ni.terminus_class = tc
         end
+      else
+        ni.terminus_class = :plain
+        node = ni.find(node, facts: facts, environment: Puppet[:environment])
```

Comment:
It's a little inconsistent that we pass the environment name here, but other places where we call `find` above we don't. It just so happens "it all magically works" because `Puppet[:environment]` is pushed onto the context by `ApplicationSupport` https://github.com/puppetlabs/puppet/blob/2ffa729d86888ae6c03eb663659357065a539967/lib/puppet/application_support.rb#L39

And then later the indirector looks up the `current environment` https://github.com/puppetlabs/puppet/blob/2ffa729d86888ae6c03eb663659357065a539967/lib/puppet/indirector/request.rb#L33

That said I think it'd be more explicit if we always passed `Puppet[:environment]` when calling `Puppet::Indirector::Indirection#find`.